packages/scripts/src/codegen/icon-codegen.ts

```diff
@@ -77,37 +91,48 @@ async function generateIcons() {
         }>
     > = Object.create(null)
 
+    const sourceMap = new Map<string, string | null>()
     await Promise.all(
         filePaths.map(async (path) => {
-            const parsedPath = parsePath(path)
+            const { ext } = parsePath(path)
+            const code = ext.toLowerCase() === '.svg' ? await readFile(path, 'utf8') : ''
+            const isDynamicColor = code ? dynamicColorRe.test(code) : false
+            const source = isDynamicColor ? svg2jsx(code) : null
+            if (source) sourceMap.set(path, source)
+        }),
+    )
 
-            const base = getBase(parsedPath.name)
-            const currentVariant = getVariant(parsedPath.name)
-            variants[base] ??= []
+    // Process in order of files
+    filePaths.forEach((path) => {
+        const { name } = parsePath(path)
 
-            // cross platform, use URL to calculate relative path
-            const importPath = './' + new URL(path, ROOT_PATH).toString().slice(relativePrefix)
-            const identifier = snakeCase(parsedPath.name)
+        const base = getBase(name)
+        const currentVariant = getVariant(name)
+        variants[base] ??= []
 
-            const url = ` /*#__PURE__*/ new URL(${JSON.stringify(importPath)}, import.meta.url)`
-            asURL.js.push(`export const ${identifier}_url = ${url}`)
+        // cross platform, use URL to calculate relative path
+        const importPath = './' + new URL(path, ROOT_PATH).toString().slice(relativePrefix)
+        const identifier = snakeCase(name)
 
-            let currentLine = `/** ${createImage(importPath)} */ export const `
-            asURL.dtsMap.addMapping({
-                generated: { line: asURL.js.length + 1, column: currentLine.length },
-                original: voidMapping,
-                source: importPath,
-            })
-            currentLine += `${identifier}_url: string`
-            asURL.dts.push(currentLine)
+        const url = ` /*#__PURE__*/ new URL(${JSON.stringify(importPath)}, import.meta.url)`
+        asURL.js.push(`export const ${identifier}_url = ${url}`)
 
-            const source = parsedPath.ext.toLowerCase() === '.svg' ? await readFile(path, 'utf8').then(svg2jsx) : null
-            variants[base].push({
-                args: [currentVariant, url, source, !!source?.match(currentColorRe)],
-                assetPath: importPath,
-            })
-        }),
-    )
+        let currentLine = `/** ${createImage(importPath)} */ export const `
+        asURL.dtsMap.addMapping({
+            generated: { line: asURL.js.length + 1, column: currentLine.length },
+            original: voidMapping,
+            source: importPath,
+        })
+        currentLine += `${identifier}_url: string`
+        asURL.dts.push(currentLine)
+
+        const isDynamicColor = sourceMap.has(path)
+        const source = isDynamicColor ? sourceMap.get(path)! : null
```

Comment:
Less code generated in favor of this checking (see https://github.com/DimensionDev/Maskbook/pull/9439/commits/b161953708751dc054ed1a6c6a578c9e0e49b06b)